src/components/heatmap/HeatmapSubscriber.js

```diff
@@ -1,125 +1,156 @@
+import React, { useEffect, useState, useCallback } from 'react';
 import PubSub from 'pubsub-js';
+import uuidv4 from 'uuid/v4';
 
 import TitleInfo from '../TitleInfo';
 import {
-  CELLS_COLOR,
-  CLUSTERS_ADD,
-  CELLS_ADD,
-  CELLS_SELECTION,
-  CLEAR_PLEASE_WAIT,
-  CELLS_HOVER,
-  STATUS_INFO,
-  CELL_SETS_VIEW,
-  RESET,
+  CELLS_COLOR, CELLS_ADD, CELLS_SELECTION,
+  CLEAR_PLEASE_WAIT, CELLS_HOVER, STATUS_INFO, CELL_SETS_VIEW,
+  RESET, EXPRESSION_MATRIX_ADD, VIEW_INFO, GENES_HOVER,
 } from '../../events';
+import { useDeckCanvasSize, pluralize, capitalize } from '../utils';
 import Heatmap from './Heatmap';
+import HeatmapTooltipSubscriber from './HeatmapTooltipSubscriber';
 
 export default function HeatmapSubscriber(props) {
   const {
-    children,
-    uuid,
-    removeGridComponent,
-    onReady,
-    variablesLabelOverride,
-    observationsLabelOverride,
-    theme,
+    removeGridComponent, onReady, theme, transpose,
+    observationsLabelOverride: observationsLabel = 'cell',
+    variablesLabelOverride: variablesLabel = 'gene',
+    disableTooltip = false,
   } = props;
-  const [cells, setCells] = useState({});
-  const [clusters, setClusters] = useState({});
-  const [cellColors, setCellColors] = useState({});
-  const [selectedCellIds, setSelectedCellIds] = useState(new Set([]));
+
+  const observationsTitle = capitalize(pluralize(observationsLabel));
+  const variablesTitle = capitalize(pluralize(variablesLabel));
+
+
+  // Create a UUID so that hover events
+  // know from which element they were generated.
+  const uuid = uuidv4();
+
+  const [cells, setCells] = useState();
+  const [expressionMatrix, setExpressionMatrix] = useState();
+  const [selectedCellIds, setSelectedCellIds] = useState(new Set());
+  const [cellColors, setCellColors] = useState(null);
   const [urls, setUrls] = useState([]);
 
   const onReadyCallback = useCallback(onReady, []);
 
+  const [width, height, deckRef] = useDeckCanvasSize();
+
   useEffect(() => {
-    function clustersAddSubscriber(msg, { data, url }) {
-      setClusters(data);
-      setUrls((prevUrls) => {
-        const newUrls = [...prevUrls].concat({ url, name: 'Clusters' });
-        return newUrls;
-      });
-    }
-    function cellsAddSubscriber(msg, { data, url }) {
-      setCells(data);
-      setUrls((prevUrls) => {
-        const newUrls = [...prevUrls].concat({ url, name: 'Cells' });
-        return newUrls;
-      });
-    }
-    function cellsSelectionSubscriber(msg, cellIds) {
-      setSelectedCellIds(cellIds);
-    }
-    function cellsColorSubscriber(msg, newCellColors) {
-      setCellColors(newCellColors);
-    }
-    const clustersAddToken = PubSub.subscribe(
-      CLUSTERS_ADD, clustersAddSubscriber,
+    const expressionMatrixAddToken = PubSub.subscribe(
+      EXPRESSION_MATRIX_ADD, (msg, { data }) => {
+        const [attrs, arr] = data;
+
+        // Get the full zarr array (all chunks & flat).
+        arr.getRaw([null, null]).then((X) => {
```

Comment:
I believe it's because this function gets called multiple times. If you look at the chunks requested I bet they are the same chunks. My suggestion (which is now added) is to fetch all the data once (like `getRaster` in viv), and then cache that data for use in the heatmap, rather than requesting the data each time it is needed each time from the source.

Reply:
I did not specify the `chunks=` argument in python, to allow Zarr to choose the best value for me. In theory it could all be one chunk since I want to request it all at the same time, but not sure if that would be worse or better in terms of networking / does HTTP2 allow multiple concurrent requests, which would make this chunking beneficial?

Reply:
Some are repeated but I'm still confused why there are chunks (i.e more than one) at all given the above.  Clearly I am missing something.
<img width="1680" alt="Screen Shot 2020-07-27 at 1 17 01 PM" src="https://user-images.githubusercontent.com/43999641/88571355-89014180-d00b-11ea-9120-dce48f89b461.png">

Reply:
@keller-mark Sorry, I should have caught that in the other PR.  I confused `shape` and `chunks`.  The zarr.js library still has to spend time assembling the chunks/your code has to do some assembling as well so I wonder if there is a way to optimize this to prevent unecessary array iteration.

Reply:
@manzt correct me if I have that wrong.

Reply:
I did not realize the requests were being made twice, but I am calling `.getRaw` twice (once in the GenesSubscriber, once in the HeatmapSubscriber), and I think this makes new requests each time. I do ultimately want two separate ArrayBuffers, but it would probably be faster to memoize and then copy the data rather than make the requests 2x.

Reply:
Zarr.js fills a single buffer concurrently, which means requests are non-blocking, but I think in the case where you are certain you want to load all the data making `chunks === shape` makes the most sense and would be the most performant. 

The real benefit that zarr provides is a standard way of representing ND arrays. This means that the same data that backs our visualization can be downloaded and used in `zarr-python` for some other type of analysis. Chunking by the shape might make sense to optimize for this particular use case, but that means the source isn't optimized for efficient random access other than "give me all the data".

Reply:
Oh I see now Trevor's comment

Reply:
I think the question to ask in this situation is "will this data be used for something other than vitessce?", if the answer is yes, than chunking as recommended by zarr probably makes sense. If not, just make one big blob because our demo _might_ be a little faster. 

In the end, loading all the data like this happens once and zarr doesn't "care" how data are chunked. So the most simple thing to do is not worry about chunking (because I doubt it makes much of a difference).

Reply:
Ah yes @keller-mark, that's what I was trying to suggest in loader. Memoize the return of `getRaw` and then return copies of the `RawArray.data` when calling `get`.

Reply:
FWIW, we do this in viv in the `StaticImageLayer`. `ZarrLoader.getRaster` wraps a call to `getRaw`, and we use React state to hold onto that data until a new `loader` or `loaderSelection` are provided:

https://github.com/hubmapconsortium/vitessce-image-viewer/blob/master/src/layers/StaticImageLayer.js

Reply:
Resolved in https://github.com/hubmapconsortium/vitessce/pull/689/commits/de062d7bde0f8ae86f908ac3b31de9e59b1156a6